pkg/controller/elasticsearch/remotecluster/remoteca/controller.go

```diff
@@ -0,0 +1,474 @@
+// It copies the remote CA Secrets so they can be trusted by every peer Elasticsearch clusters.
+func (r *ReconcileRemoteCa) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	defer common.LogReconciliationRun(log, request, "es_name", &r.iteration)()
+	tx, ctx := tracing.NewTransaction(r.Tracer, request.NamespacedName, "remoteca")
+	defer tracing.EndTransaction(tx)
+
+	// Fetch the local Elasticsearch spec
+	es := esv1.Elasticsearch{}
+	err := r.Get(request.NamespacedName, &es)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return deleteAllRemoteCa(ctx, r, request.NamespacedName)
+		}
+		return reconcile.Result{}, err
+	}
+
+	if common.IsPaused(es.ObjectMeta) {
+		log.Info("Object is paused. Skipping reconciliation", "namespace", es.Namespace, "es_name", es.Name)
+		return common.PauseRequeue, nil
+	}
+
+	enabled, err := r.licenseChecker.EnterpriseFeaturesEnabled()
+	if err != nil {
+		return defaultRequeue, err
+	}
+	if !enabled {
+		log.Info(
+			"Remote cluster controller is an enterprise feature. Enterprise features are disabled",
+			"namespace", es.Namespace, "es_name", es.Name,
+		)
+		return reconcile.Result{}, nil
+	}
+
+	// Use the driver to create the remote cluster
+	return doReconcile(ctx, r, &es)
+}
+
+// deleteAllRemoteCa deletes all associated remote certificate authorities
+func deleteAllRemoteCa(ctx context.Context, r *ReconcileRemoteCa, es types.NamespacedName) (reconcile.Result, error) {
+	span, _ := apm.StartSpan(ctx, "delete_all_remote_ca", tracing.SpanTypeApp)
+	defer span.End()
+
+	currentRemoteClusters, err := getCurrentRemoteCertificateAuthorities(ctx, r.Client, es)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	results := &reconciler.Results{}
+	for remoteCluster := range currentRemoteClusters {
+		if err := deleteCertificateAuthorities(ctx, r, es, remoteCluster); err != nil {
+			results.WithError(err)
+		}
+	}
+	return results.Aggregate()
+}
+
+func doReconcile(
+	ctx context.Context,
+	r *ReconcileRemoteCa,
+	localEs *esv1.Elasticsearch,
+) (reconcile.Result, error) {
+	localClusterKey := k8s.ExtractNamespacedName(localEs)
+	// Get current clusters according to the existing remote CAs
+	currentRemoteCertificateAuthorities, err := getCurrentRemoteCertificateAuthorities(ctx, r.Client, localClusterKey)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	expectedRemoteCertificateAuthorities, err := getExpectedRemoteCertificateAuthorities(ctx, r.Client, localEs)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	results := &reconciler.Results{}
+	// Create or update expected remote CA
+	for remoteEsKey := range expectedRemoteCertificateAuthorities {
+		// Get the remote Elasticsearch cluster associated with this remote CA
+		remoteEs := &esv1.Elasticsearch{}
+		if err := r.Client.Get(remoteEsKey, remoteEs); err != nil {
+			if !errors.IsNotFound(err) {
+				return reconcile.Result{}, err
+			}
+			// Remote cluster does not exist, skip it
+			continue
```

Comment:
The other way around reads IMHO better because of the dual negation (not is not found):
```go
if errors.IsNotFound(err) {
    // Remote cluster does not exist, skip it
    continue
}
return reconcile.Result{}, err
```